### DISCUSSION TITLE
how to keep z axis rotate in a range?

### DISCUSSION THREAD

**Discussion by vegfood** [tesseract-robotics/tesseract_planning#239] [Q&A]
Problem: I want to use the ur robot to hold a glass of water and move it from a place to another, in the process i don't want the water spit , so the z axis can't rotate too much, is it possible to achieve this in tesseract? I think a constraint is needed to keep the z aixs rotate in a range , I tried the cartesian_coeff but i find for a trajectory(start pos and end pos are both cartesian wp), if i set the first three coeff to zero of cartesian_coeff(which i thought would only constrain the rotation of end_frame), the robot will not move to my target goal, only rotate to the desied orientation

**Comment by Levi-Armstrong** [comment]
> I want to use the ur robot to hold a glass of water and move it from a place to another, in the process i don't want the water spit , so the z axis can't rotate too much, is it possible to achieve this in tesseract? 

Yes, you define an orientation constraint. Take a look at the glass up right example. 

Also I believe the example should provide you what you need for the problem you described.

**Reply by vegfood** [reply]
you mean the glass_upright_example.cpp or the glass_upright_ompl_example.cpp ? Because in the former script, it only uses the cartesian_coeff which doesn't work in my situation.

**Reply by Levi-Armstrong** [reply]
Why will the Cartesian constraint not work?

**Reply by vegfood** [reply]
I tried to move from A to B(A,B are both waypoints), I set the cartesian_coeff as [0 0 0 5 5 5] becaues I don't want the orientation changed  during the movement, but I find if I set the coeff like this,  my robot will not move at all, it will only rotate to my target orientation.

**Reply by Levi-Armstrong** [reply]
Is the start and end state using the same Cartesian constraint or are they joint waypoints?

**Reply by Levi-Armstrong** [reply]
Also if you are able to post your Composite Instruction and profiles that should be enough for me to identify the issue.

**Reply by vegfood** [reply]
CompositeInstruction program(
            "cartesian_program", CompositeInstructionOrder::ORDERED,
            ManipulatorInfo("manipulator", "base_link", "tool0"));

    PlanInstruction start_instruction(start_point, PlanInstructionType::START);
    program.setStartInstruction(start_instruction);

    PlanInstruction plan_f0(wp1, PlanInstructionType::FREESPACE, "freespace_profile");
    plan_f0.setDescription("from_start_plan");
    PlanInstruction plan_c0(wp2, PlanInstructionType::FREESPACE, "RASTER");
    program.push_back(plan_f0);
    program.push_back(plan_c0);

composite_profile->collision_cost_config->type = tesseract_collision::CollisionEvaluatorType::CONTINUOUS;
composite_profile->collision_cost_config->contact_manager_config = tesseract_collision::ContactManagerConfig(0.00);
composite_profile->collision_cost_config->collision_margin_buffer = 0.03;
composite_profile->collision_cost_config->collision_coeff_data = 30;

composite_profile->collision_constraint_config->type = tesseract_collision::CollisionEvaluatorType::CONTINUOUS;
composite_profile->collision_constraint_config->contact_manager_config = tesseract_collision::ContactManagerConfig(0.001);
composite_profile->collision_constraint_config->collision_margin_buffer = 0.02;
composite_profile->collision_constraint_config->collision_coeff_data = 0;

auto plan_profile1 = std::make_shared<TrajOptIfoptDefaultPlanProfile>();
plan_profile1->cartesian_coeff = Eigen::VectorXd::Constant(6,1,1);

auto plan_profile2 = std::make_shared<TrajOptIfoptDefaultPlanProfile>();
plan_profile2->cartesian_coeff = Eigen::VectorXd::Constant(6,1,5);
plan_profile2->cartesian_coeff(1) = 0;
plan_profile2->cartesian_coeff(2) = 0;
plan_profile2->cartesian_coeff(3) = 0;

planning_server.getProfiles()->addProfile<TrajOptIfoptPlanProfile>(
        profile_ns::TRAJOPT_IFOPT_DEFAULT_NAMESPACE, "freespace_profile", plan_profile1);
planning_server.getProfiles()->addProfile<TrajOptIfoptPlanProfile>(
        profile_ns::TRAJOPT_IFOPT_DEFAULT_NAMESPACE, "RASTER", plan_profile2);

**Reply by vegfood** [reply]
wp1 and wp2 are two cartesian waypoints, i want to keep z axis always up while moving from wp1 to wp2

**Reply by Levi-Armstrong** [reply]
Okay so the issue is that the waypoint needs to have one set of constraints and the path between them need to have another set of constraints. So a plan instruction has a profile and path profile and if a path profile is not defined it used the profile for the path. The profile is associated with the waypoint assigned to the plan instruction and the path profile is assigned to how you move to the waypoint. In your case, you need to create two profiles one which has all of the coefficients set to some value for the "profile" then another which has the following coefficients [0,0,0,5,5,0] for the "path profile".

**Reply by vegfood** [reply]
Is there any example of setting a path profile ?

**Reply by vegfood** [reply]
PlanInstruction plan_c0(wp2, PlanInstructionType::FREESPACE, "RASTER","Z_UP");
        auto path_profile = std::make_shared<TrajOptIfoptDefaultPlanProfile>();
        path_profile->cartesian_coeff = Eigen::VectorXd::Constant(6,1,5);
        path_profile->cartesian_coeff(0) = 0;
        path_profile->cartesian_coeff(1) = 0;
        path_profile->cartesian_coeff(2) = 0;
        path_profile->cartesian_coeff(5) = 0;
        planning_server.getProfiles()->addProfile<TrajOptIfoptPlanProfile>(
                profile_ns::TRAJOPT_IFOPT_DEFAULT_NAMESPACE, "Z_UP", path_profile);

**Reply by vegfood** [reply]
Is this the correct way ?

**Reply by Levi-Armstrong** [reply]
Use this method to assign it to the plan instruction.

https://github.com/tesseract-robotics/tesseract_planning/blob/45433bace4e851ac8bf2c93910020f218ab0cf1d/tesseract_command_language/include/tesseract_command_language/plan_instruction.h#L107-L108

**Reply by vegfood** [reply]
PlanInstruction plan_c0(wp2, PlanInstructionType::LINEAR, "RASTER");
    plan_c0.setPathProfile("Z_UP");

   auto path_profile = std::make_shared<TrajOptIfoptDefaultPlanProfile>();
   path_profile->cartesian_coeff = Eigen::VectorXd::Constant(6,1,5);
   path_profile->cartesian_coeff(0) = 0;
   path_profile->cartesian_coeff(1) = 0;
   path_profile->cartesian_coeff(2) = 0;

   planning_server.getProfiles()->addProfile<TrajOptIfoptPlanProfile>(
           profile_ns::TRAJOPT_IFOPT_DEFAULT_NAMESPACE, "Z_UP", path_profile);

**Reply by vegfood** [reply]
Is this the correct way?

**Reply by Levi-Armstrong** [reply]
I believe so.

**Reply by vegfood** [reply]
is it possible to allow the z axis rotate in a range, not just fix this axis. I notice there is constraint_error_functions in the plan_profile, i wander if i have to define my own error_function to achieve this ?

**Reply by Levi-Armstrong** [reply]
This currently not support out of the box but you could do as you suggested by implementing your own constraint_error_functions.
